### PR TITLE
Add missing newlines in hover tooltips

### LINKED ISSUE #10511: Missing newline or whitespace in variable hover popover
### Check for existing issues

- [X] Completed

### Describe the bug / provide steps to reproduce it

When hovering a symbol, in the `variable x` context popover, after the `provided by "header.h"` text, there appears to be a missing newline or whitespace. Right now, the `provided by "header.h"` text runs directly into the following `Type: ` section, which looks like it should be on the next line, to align with the `Value = ` text.

Right now, you will have something like:

```
variable x

provided by "header.h"Type: uint32_t
Value = 123
```

Which should probably be:

```
variable x

provided by "header.h"
Type: uint32_t
Value = 123
```

Also, minor note, but the capitalization, punctuation, and spacing of these elements is also inconsistent. `Type: ` is capitalized and using a colon, while `variable` and `provided by` are not capitalized and have no punctuation, and `Value = ` is capitalized with an equals sign, and there appears to be an extra empty line after `Value =`. If these were all normalized:

```
Variable: x
Provided by: "header.h"
Type: uint32_t
Value: 123

// Definition
```

### Environment

Zed: v0.130.6 (Zed)
OS: macOS 11.7.2
Memory: 16 GiB
Architecture: aarch64

### If applicable, add mockups / screenshots to help explain present your vision of the feature

zed-hover-popover

```
Variable: tty_uart

Provided by: "uart.h"
Type: 'UART_HandleTypeDef (aka struct __UART_HandleTypeDef)'
Value: '&tty_uart'

extern UART_HandleTypeDef tty_uart
```

or

```
variable tty_uart

provided by "uart.h"
Type: 'UART_HandleTypeDef (aka struct __UART_HandleTypeDef)'
Value = '&tty_uart'

extern UART_HandleTypeDef tty_uart

```

### If applicable, attach your `~/Library/Logs/Zed/Zed.log` file to this issue.

_No response_

### PR DESCRIPTION
**Pull request by CharlesChen0823**
fix #10511 

Release Notes:
- N/A

### CODE REVIEW

File: crates/language/src/markdown.rs [hunk 1]
```diff
@@ -301,6 +301,7 @@ pub async fn parse_markdown_block(
                 TagEnd::Heading(_) => bold_depth -= 1,
                 TagEnd::CodeBlock => current_language = None,
                 TagEnd::Emphasis => italic_depth -= 1,
+                TagEnd::Paragraph => text.push('\n'),
```
[the remark this answers is not in the document]
Comment (CharlesChen0823) [reply crates/language/src/markdown.rs:304]: sorry, i made a mistake. missing line is caused by the lake of metadatablock.

**Comment by CharlesChen0823** [comment]
@mrnugget when enter metadata block, start new parser for content in block. if this is an acceptable method?

**Comment by mrnugget** [comment]
> @mrnugget when enter metadata block, start new parser for content in block. Is this an acceptable method?

Hmmm, I'm a bit skeptical of recursion here. Do you have a specific bit of Markdown to reproduce the issue? I think we at least need that.

**Comment by CharlesChen0823** [comment]
```
2024-04-16T15:58:00+08:00 TRACE lsp] incoming message: {"id":4,"jsonrpc":"2.0","result":{"contents":{"kind":"markdown","value":"### function `search`  \n\n---\n→ `void`  \nParameters:  \n- `TrieNode * flag (aka struct TrieNode *)`\n- `char * word`\n\n---\n```cpp\nvoid search(TrieNode *flag, char *word)\n```"},"range":{"end":{"character":8,"line":230},"start":{"character":2,"line":230}}}}

```

**Comment by mrnugget** [comment]
Okay, so the markdown looks like this:

````
### function `search`

---
→ `void`
Parameters:
- `TrieNode * flag (aka struct TrieNode *)`
- `char * word`

---
```cpp
void search(TrieNode *flag, char *word)
```
````

That makes me wonder: aren't the `---` just supposed to be dividers? Like, if I put a `---` in the next line here

---

then it renders as a `<hr>` I think. So should we handle that?

**Comment by bennetbo** [comment]
> Okay, so the markdown looks like this:
> 
> ```
> ### function `search`
> 
> ---
> → `void`
> Parameters:
> - `TrieNode * flag (aka struct TrieNode *)`
> - `char * word`
> 
> ---
> ```cpp
> void search(TrieNode *flag, char *word)
> ```
> ```
> 
> That makes me wonder: aren't the `---` just supposed to be dividers? Like, if I put a `---` in the next line here
> 
> then it renders as a `<hr>` I think. So should we handle that?

Hmm I think there are two problems which we will run into when rendering dividers:
- It looks like our markdown parser dependency (pulldown_cmark) does not pickup dividers if there is no line break before and after the `---`, see https://www.markdownguide.org/basic-syntax/#horizontal-rule-best-practices. Github is able to render that though
- The markdown crate renders everything into an `InteractiveText` element, which does not support inserting other elements then text. We could 'split' up blocks [here](https://github.com/zed-industries/zed/blob/db73bb9e3080205b018ed38fd31b52cd83e93080/crates/editor/src/editor.rs#L152) and render dividers, however at this point it would probably make the most sense to combine the `language` markdown implementation with the `markdown_preview` markdown implementation (which does support lists, tables, dividers, etc.)

**Comment by CharlesChen0823** [comment]
In markdown, start `---` and end with `---` is an extentsion named markdown yaml metadata block, and there also can be start `+++` with end `+++`.  this is not dividers

**Comment by mrnugget** [comment]
> In markdown, start `---` and end with `---` is an extentsion named markdown yaml metadata block

Isn't that YAML frontmatter? YAML frontmatter is supposed to only show up at the start of a text (see [here](https://jekyllrb.com/docs/front-matter/)) and "must take the form of valid YAML set between triple-dashed lines".

In this case it doesn't look like that's YAML between the `---`.

**Comment by mrnugget** [comment]
Neovim also renders it as dividers: 

screenshot-2024-04-16-12 16 52@2x

With this being the Markdown:

````
### instance-method `format`

---

→ `void`
Parameters:
- `const int &`
- `const std::tm &`
- `int & dest`

---

```cpp
// In my_formatter_flag
public: void format(const int &, const std::tm &, int &dest)
```
````

**Comment by mrnugget** [comment]
Closing this in favor of https://github.com/zed-industries/zed/pull/10606

**Pull request closed without merging**